### ISSUE TITLE
How should sync and single variables be read/written (in terms of I/O)

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#15891]
At present, trying to read() or write() a sync/single variable results in a compiler error because the last time we discussed how they should be handled, we couldn't come up with an answer that was satisfactory.  However, it seems unfortunate that they are among a very small number of types that can't be read/written, and this property tends to spill into other uses of them, sometimes in subtle and indirect ways as noted in issue #14467.

This question asks how they should be read and written.  I'll start with writing them since that's more common, though reading them is where things get tricky, I think.

Options for writing syncs
-------------------------

* **readXX stye**: print out the value, leaving the full/empty state unchanged (this would act like a "peek" into the variable without interacting with it and ignores the full/empty state)
* **readFE style**: print out the value, blocking until full and leaving empty (this would act the most like any other undecorated "use" of the sync variable)
* **state style**: print out something like `<empty>` if the variable is empty, and print its value if it's full (this would be the most like capturing the full state of the variable, and of course a user could always apply an explicit `.readXY()` method to the sync if they wanted something different
* **other?**

Options for reading syncs
--------------------------
* **writeEF style**: read in a value, blocking until empty, leaving full (this would act the most like any other undecorated "definition" of the sync variable)
* **overwrite style**: regardless of the current state of the variable, store the value into the sync leaving it full, or if it is the special value "<empty>" (or whatever), make the variable empty
* **writeXF style**: always read in a value, and store it into the variable whatever the sync's state is

Note that historically, I believe we used "readFE style" for write() and "writeEF style" for read() but that this ended up being confusing to users because they'd put in a debug writeln() for their program, end up altering the full/empty state, and cause their program to deadlock.

For writing, my preference is "state style" because it seems to capture the true nature of the sync.  However, from a symmetry perspective, that suggests using "overwrite style" for read() which seems a bit odd to me, intuitively.  For reading, "writeEF style" seems most natural to me, yet that suggests using "readFE style" for writing, which I don't like (and returns us to where we were).

**Comment by mppf** [comment]
I was thinking we'd make a write(ln) use readXX and a read(ln) on a sync result in compilation error.

The cases that came up in #14467 and in @ronawho's PR come from a "Any chapel type can be printed" viewpoint.

Does your *state style* suggestion for write(ln) involve changing the full/empty state of the sync variable, or is it more "peeking" like readXX? I'd be happy with that if it were "peeking".

If we didn't want reading sync variables to be an error - I think it'd depend on what happens when writing. I keep trying to advocate for something printed with write(ln) to be readable with read(ln) (see #5052 and #14107) and so I would want the read function to handle the format emitted by the write function.

**Comment by bradcray** [comment]
> Does your state style suggestion for write(ln) involve changing the full/empty state of the sync variable, or is it more "peeking" like readXX? I'd be happy with that if it were "peeking".

Right, the intention is to have it print an indication of its state without changing it:  It's either empty (in which case it prints out some sort of sentinel string) or it's full (in which case it prints out the value).

Your "printed with writeln() should be read with readln()" philosophy is what led me to suggest the overwrite style.  It seems like the natural complement to the _state_ style.  Though, like some other types/cases (e.g., strings with whitespace), it's imperfect without additional formatting/meta-information.  For example, if we used "<empty>" for the sentinel string for an empty sync var and you had a `sync string` that was full but stored the value `<empty>`, you'd be in trouble.  Of course, like strings, we could print out additional information to help distinguish between those cases (e.g., all sync values are written out as `sync(<empty>)` or `sync([value goes here])`), but that doesn't seem as nice as a default print option (similar to how having strings print with quotes to demarcate them probably isn't what people want by default).

I was thinking we wouldn't want reading syncs to result in a compilation error because it seems like we'd get into the same kinds of problems as #14467 (assuming that we took the same implementation approach):  As soon as the compiler needed to resolve a dynamic dispatch on `readThis()` in a program that contained a code path that led to `sync.readThis()` we'd get a compiler error.

**Comment by cassella** [comment]
I have a few thoughts:

1. This issue is asking "How should a sync variable be written?".  Does that have to have the same answer as "How should a record containing a sync variable write that sync variable?"

2. I'd like to make an argument for `writeln(asyncvar$)` continue to be forbidden, or use `readFE()`, but `writeln(r)` using `state style`, or `readXX()`, for a sync variable the record `r` contains.

3. I think this should also be considered in the default hashing function for a record containing a sync.

In more detail,

1.

I think writing a sync variable and writing a record containing a
sync variable are different.  In the former case, the user should know
the thing they're operating on is a sync variable, and they have no
way to alter the default behavior of the `syncVar.readWritethis()`.

In the latter case, the user is writing e.g. `writeln(r)` as a
convenience over writing each individual field, and they can't special
case the sync field in the middle.  In this case, the decision in this
issue is only setting the behavior of the compiler-generated
`R.readWriteThis()`.  If the record author can override it with
something more appropriate if needed.  (Though they can't easily
change the treatment of one sync variable in a record of hundreds of
fields.)


2.

I agree with the expectation of that debugging writeln() case, that
generally one would expect that writeln() wouldn't modify the state of
the thing being written. (aside from taking locks, etc.)  So I think
`state mode` is a reasonble default for record members.

Continuing to forbid a bare `writeln(syncvar$)` seems reasonable, to
force the user to spell out what they mean.  

For the record member case, I think `overwrite` may be a better match
for `readFE` than `writeEF` is.  I normally think of reading a record as
completely overwriting the old value -- almost deinitializing the record
and reinitializing it from the read data.  (And again, the record
author can override the behavior if `writeEF` is a better match.)


3.

I think the default hash function for a record containing a sync
variable should use `state style`.  Even moreso than for writing, it
doesn't make sense that a hash function would change the value being
hashed.  If `set.add(r)` consumes the sync variables in `r`, the
object stored in the set no longer hashes to the slot it was put in.

It makes a little more sense here to require the record author to
spell out what they want, but I think `state style` is more likely to
be what they want.

Today, this hangs after printing "added":

```chapel
import Set;

record R {var f$: sync int; proc init(v: int) { f$ = v;} }
var s = new Set.set(R);

var r = new R(5);
s.add(r);

writeln("added");

for rr in s {
  writeln("in loop");
  writeln("got an ", rr.f$.readXX());
}
```

(This may just be a problem with `Set`.)

**Comment by bradcray** [comment]
I understand your argument that a record author has a mechanism for changing how the value printed, but a sync var author doesn't.  But I don't see why that would suggest not using the same rule for the record field as the standalone case for consistency/simplicity.  Even if there's a default, the user can still apply an explicit method to override that default in the standalone case.  And it seems generally annoying to have just one type (I think?) that has no default way of printing itself out (breaks generic code, for example).

In fact, if I were using your argument, I might suggest that record fields shouldn't have a default way of printing them out (because the record author can and should have indicated how it should print) whereas standalone variables don't have a way to override (so will remain the lone unprintable type, which I think is unattractive).  (Really, though, I want them to be the same, I'm just playing devil's advocate here).

I think an advantage to writing out "state style" in both cases is that if you got a different behavior than you expected ("oh, that printed empty when I expected it to print a value"), it's a cue that you need to update your code.  Whereas readXX style can print an arbitrary value without you knowing whether it was properly set or not (i.e. if you believed the semantics would be readEF, nothing will make you aware of your mistake, and you may rely on behavior that isn't there).

(This also, conveniently, matches your overwrite style for reading...)

**Comment by vasslitvinov** [comment]
Intuitively, I prefer state style for writing and disallow for reading. In which case this can be the same behavior for sync variables and sync-containing records. Somehow I do not expect good things to come out of overwrite style for reading.

I am OK with asymmetric treatment of writing and reading, given symmetric alternatives have plenty of issues.

Maybe we can do something fancier for reading? Ex.:
* block until the variable is empty in any case
  this implies that reading into a full single variable is a deadlock; I believe it is important to preserve the single-ness invariant
* if the value being read is sync(<empty>), leave the variable empty, otherwise fill the variable with the value being read

I think this fancier reading option blends well with "state style" writing, like sync(value|empty). Maybe for an empty sync string variable we can distinguish it from sync(), ex. sync<>.

**Comment by vasslitvinov** [comment]
Having the user override the default read/write behavior of a sync field by defining read/writeThis on the enclosing record is an option, yet it is less than ideal. Because this means writing a lot of boilerplate and keeping it up to date with changes in record fields.

To mitigate that, perhaps hint at using field-erator with special handling of field(s) of desired name(s).

**Comment by ronawho** [comment]
Is there a current best practice for creating a class with a sync?

I'm trying to create a boxed sync like:

```chpl
class BoxedSync {
  type T;
  var s: sync T;
  forwarding s;
}
```

for internal usage and I get the error about not being able to write syncs. Should I create a no-op `writeThis` (are there others I need to create too?), use `pragma "no default functions"`, or something else?

**Comment by ronawho** [comment]
Creating a `writeThis` seems to work for the tests that were failing, but I can never remember how these functions work, so wasn't sure if I also needed a `readWriteThis` overload or any others.

**Comment by bradcray** [comment]
IIRC, even the writeThis _shouldn't_ be necessary if you're not printing out variables of this type, but that it may end up getting required because of the fact that we try to create a global dynamic dispatch table for all classes and resolving it if any class is potentially written out in the program (?).  So it may be that if the program contained any read() calls of a class, that you might need readThis as well (?).

**Comment by mppf** [comment]
> Creating a `writeThis` seems to work for the tests that were failing, but I can never remember how these functions work, so wasn't sure if I also needed a `readWriteThis` overload or any others.

Yeah you just need `writeThis`, and that should disable generation of `readThis`, and do a reasonable thing for `readWriteThis`.

**Comment by mppf** [comment]
In the encoder/decoder proposal (#18499) there is a desire to have the read operation produce a new value and then, if necessary, that can be assigned. That means that `sync` might not have a `readThis` at all; instead it might have a special initializer `proc sync.init(r: fileReader)` or maybe a factory function `proc sync.createByReading(r: fileReader)`. Either way, the sync object will be newly constructed for the read and so there is no concern about race conditions since nothing else can access that newly allocated memory.

If it is an error to assign sync variables, we might still get compilation errors if you do `read(mySyncVar)`, but that error won't involve dynamic dispatch (it would be more just "You can't assign sync variables").

**Comment by bradcray** [comment]
> we might still get compilation errors if you do read(mySyncVar), but that error won't involve dynamic dispatch (it would be more just "You can't assign sync variables"). 

I think this _may_ actually still result in the dynamic dispatch errors.  Specifically, the way this happens today is:
* user defines a class with field that has a sync type
* compiler generates a default read method for that class that reads each of its fields
* user reads some other class instance that is statically indistinguishable from the one in question
* dynamic dispatch resolves the case with the error and generates the error

Unless the compiler-generated read method was also an init-based reader, in which case maybe all is good (?).

**Comment by mppf** [comment]
> Unless the compiler-generated read method was also an init-based reader, in which case maybe all is good (?).

Yes, that's exactly what I would expect to happen.
